endpoints/appr/cnr_backend.py

```diff
@@ -12,7 +17,97 @@
 from util.request import get_request_ip
 
 
+class LazyPackage(object):
+    def __init__(self, blob=None, b64_encoded=True):
+        self.files = {}
+        self.tar = None
+        self.blob = None
+        self._io_file = None
+        self._digest = None
+        self._size = None
+        self.b64blob = None
+        if blob is not None:
+            self.load(blob, b64_encoded)
+
+    def _load_blob(self, blob, b64_encoded):
+        if b64_encoded:
+            self.b64blob = blob
+            self.blob = base64.b64decode(blob)
+        else:
+            self.b64blob = base64.b64encode(blob)
+            self.blob = blob
+
+    def load(self, blob, b64_encoded=True):
+        self._digest = None
+        self._load_blob(blob, b64_encoded)
+
+    @property
+    def io_file(self):
+        self._lazy_load_file()
+        return self._io_file
+
+    def _lazy_load_file(self):
+        if self._io_file is not None:
+            return
+
+        self._io_file = io.BytesIO(self.blob)
+
+    def _lazy_load_tar(self):
```

Comment:
Do you need any locking around these methods?

Reply:
I don't believe so, as this is only called per thread

Reply:
Its literally instantiated per request... what are your concerns around locking? (especially give the likelyhood this isn't even really used)

Reply:
Right now we're instantiating one per request, but we've also retroactively added fully hydrated object caches to hot routes in the past. This is a textbook example of when you need locking, or at least an atomic compare and swap. You've got shared state and multiple methods on the object can update that state. Moreover, something that is masquerading as a simple property read is actually mutating the object state. Why not just add it?

Reply:
Mostly a concern about additional overhead, and a belief that the "at worst" scenario is that we simply reload the data, since the attributes being filled in will have the same result even if we call the loading code twice

Reply:
Here's my general concern:

A list of heap allocated objects of this type are stashed in a per-gunicorn-worker cache. Two greenlets get the same reference returned from the cache concurrently. Now depending on where/when the greenlets context swap, you have a scenario where they could both be mutating memory in the object in a way that corrupts the state.